ml-agents/mlagents/trainers/policy/checkpoint_manager.py

```diff
@@ -0,0 +1,133 @@
+# # Unity ML-Agents Toolkit
+from typing import Union, Dict, Any, Optional
+import os
+from enum import Enum
+import attr
+from collections import defaultdict
+from mlagents.trainers.training_status import GlobalTrainingStatus
+from mlagents_envs.logging_util import get_logger
+
+logger = get_logger(__name__)
+
+
+@attr.s(auto_attribs=True)
+class Checkpoint:
+    steps: int
+    file_path: str
+    reward: Optional[float]
+    creation_time: str
+
+
+class CheckpointType(Enum):
+    CHECKPOINT = "checkpoint"
+    STEPS = "steps"
+    FINAL_PATH = "final_model_path"
+    REWARD = "reward"
+    CREATION_TIME = "creation_time"
+
+
+class CheckpointManager:
```

Comment:
Originally they were all handled in GlobalTrainingStatus, but the idea of separating them was to ensure that GlobalTrainingStatus only handles the final storage and CheckpointManager deals with the actual logic of the checkpoint additions/removals.

Reply:
Should CheckpointManager just call GlobalTrainingStatus under the hood? So that GlobalTrainingStatus remains the "general" implementation but CheckpointManager uses it as a utility to write information, as well as do checkpoint-specific stuff.

Reply:
I believe this concern should be resolved, now `CheckpointManager` is stateless and uses `GlobalTrainingStatus` for managing checkpoint state.